### ISSUE TITLE
Sourcing a file from a function breaks variable exporting

### ISSUE THREAD

**Issue by JoshCheek** [fish-shell/fish-shell#5064]
Configuration
-------------

* `fish --version` "fish, version 2.7.1"
* `echo $version` "2.7.1"
* `name -a` "Darwin Joshs-MacBook-Air.local 17.6.0 Darwin Kernel Version 17.6.0: Tue May  8 15:22:16 PDT 2018; root:xnu-4570.61.1~1/RELEASE_X86_64 x86_64"
* `echo $TERM` "xterm-256color"
* `sh -c 'env HOME=$(mktemp -d) fish'` yes, you'll see it in the asciinema at the end

Synopsis
---------

Okay. So, if a variable is not set, and I call a function which sources a file which exports that variable. Then the variable will still not be set. It will be declared local to the function. IMO, this is a bug, `man set` says _"1.  If a variable is explicitly set to either be exported or not exported, that setting will be honored."_

Reproduce
-----------

You can reproduce it with this config.fish, it calls `maybe_source`, because not all my environments share the same files.

```fish
# config.fish
set --export ALREADY_DEFINED set_in_public_config

function maybe_source
  if test -e $argv[1]
    source $argv[1]
    echo '----- maybe_source -----'
    set --show | ag -v history | ag DEFINED
  end
end
maybe_source $HOME/.config/fish/private_config.fish

echo '----- config.fish -----'
set --show | ag -v history | ag DEFINED
```

And this `private_config.fish`, which I use to set variables that I don't want to store in git in my dot files (eg they have sensitive values or the values change across the machines I work on).

```fish
# private_config.fish
set --export ALREADY_DEFINED set_in_private_config
set --export NOT_DEFINED     set_in_private_config
echo '----- private_config.fish -----'
set --show | ag -v history | ag DEFINED
```

I recorded it at https://asciinema.org/a/HjQTxey7c4JI7eG3kt5SSYrKn, as the issue template suggested. It's a little bit roundabout, and the temp home dir messed with my editor, but I didn't want to go down a rabbit hole trying to make it perfect (it's taken me a couple of hours to debug it this far).

**Comment by faho** [comment]
Unfortunately, this is not a bug. `set -x` does not specifically define a scope. That means for $NOT_DEFINED:

>If a variable is not explicitly set to be either universal, global or local and has never before been defined, the variable will be local to the currently executing function.

and for $ALREADY_DEFINED:

>If a variable is not explicitly set to be either universal, global or local, but has been previously defined, the previous variable scope is used.

If you want a global variable, define a global variable. Use `set -g`. If it also needs to be passed to child processes ("exported"), use `set -gx`.

**Comment by JoshCheek** [comment]
Okay, after thinking about it a bit, I think I can see the rationale for this, which is:

You might want to set an env var for the duration of a local scope, and then unset it when that scope ends (I think one reason this was so surprising was that I was thinking of env vars as separate from fish's local and global vars)

It feels strange to me that a file's scope will change when evaluted by `source` vs my `maybe_source`, but maybe that's just because it's a new idea to me and so I wasn't expecting it. I guess `source` is kind of like a macro in that it evaluates in the caller's scope, but `maybe_source` is like a function that gets its own scope. If the context is useful, [here](https://github.com/JoshCheek/dotfiles/commit/a62a51bb2352d10b4f8e4fdb8006bcea597618d5) is where I introduced it. The intent was to encapsulate the conditional check on chruby, nodenv, and private_config (conditional b/c not all my machines have the same setup, but they do all have the same dotfiles)

---

>       The exporting rules when creating or updating a variable are identical to the scoping rules for variables:
>
>       1.  If a variable is explicitly set to either be exported or not exported, that setting will be honored.

I think my confusion was because I considered environment variables, local variables, and global variables to be different namespaces. So in my head, exporting the variable had nothing to do with whether it was local or global, hence the behaviour I observed seemed incongruent with the man page I quoted. I think the "identical to the scoping rules" was trying to tell me this, but I did not interpret it to mean that until just now as I was asserting that this should be documented there 😝

---

Aaaaanyway:

I'll update my private config to declare my env vars as global and exported. 

I propose changing the wording on the man page to make this a little clearer (b/c I read it 2 or 3 times while making the original post, and didn't realize this). Here are two potential alternatives:

* _"a variable is exported for the duration of its scope"_
* _"a variable is unexported when it goes out of scope"_

If you're not into this proposal, then feel free to close. If you like it but feel like I'm just making work for you, then LMK, I've gotten Fish built from source before, so I should be able to make the PR.
